src/ai/backend/manager/models/image.py

```diff
@@ -895,95 +997,125 @@ async def _in_domain_scope(
-            raise InvalidScope(f"Project not found (project_id:{scope.project_id})")
+            raise InvalidScope(f"Project not found (project_id: {scope.project_id})")
+
+        return await self.calculate_permission(ctx, scope)
+
+    async def _in_project_scopes_by_registry_condition(
+        self,
+        ctx: ClientContext,
+        scopes: list[ProjectScope],
+        registry_condition_factory: Callable[[list[Any]], Any],
+        filter_global_registry: bool = False,
+        filter_customized_image: bool = False,
+    ) -> ImagePermissionContext:
+        from .container_registry import ContainerRegistryRow
+
+        if not scopes:
+            raise InvalidScope("No project scopes provided")
+
+        project_ids = [scope.project_id for scope in scopes]
+
+        project_id_to_permissions: dict[str, frozenset[ImagePermission]] = {}
+        for scope in scopes:
+            permissions = await self._verify_project_scope_and_calculate_permission(ctx, scope)
+            project_id_to_permissions[str(scope.project_id)] = permissions
+
+        registry_filter = registry_condition_factory(project_ids)
 
         image_select_stmt = (
             sa.select(ImageRow)
-            .select_from(
-                sa.join(
-                    ImageRow, ContainerRegistryRow, ImageRow.registry_id == ContainerRegistryRow.id
-                )
-            )
-            .where(
-                sa.or_(
-                    ContainerRegistryRow.is_global == true(),
-                    sa.and_(
-                        ContainerRegistryRow.is_global == false(),
-                        ContainerRegistryRow.association_container_registries_groups_rows.any(
-                            group_id=scope.project_id
-                        ),
-                    ),
+            .join(ImageRow.registry_row)
+            .options(
+                joinedload(ImageRow.registry_row).joinedload(
+                    ContainerRegistryRow.association_container_registries_groups_rows
                 )
             )
+            .where(registry_filter)
         )
 
-        for row in await self.db_session.scalars(image_select_stmt):
-            _row = cast(ImageRow, row)
-            image_id_permission_map[_row.id] = permissions
+        image_id_permission_map: dict[UUID, frozenset[ImagePermission]] = {}
+
+        result = (await self.db_session.scalars(image_select_stmt)).unique()
+        for img_row in result:
+            img_row = cast(ImageRow, img_row)
+
+            if filter_customized_image:
+                allowed_registries = await self._get_allowed_registries_for_user(ctx, ctx.user_id)
+                if not self._is_image_accessible_for_user(img_row, allowed_registries, ctx.user_id):
+                    continue
+
+            if filter_global_registry:
+                # Assumption: All global registry have same permission
+                image_id_permission_map[img_row.id] = list(project_id_to_permissions.values())[0]
```

Comment:
I think this may not be an issue right now, but it seems like code that could potentially cause problems later.